subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/WatchedHierarchies.java

```diff
@@ -16,99 +16,58 @@
+    private WatchedHierarchies(FileHierarchySet watchedHierarchies, ImmutableSet<File> watchedRoots) {
+        this.watchedRoots = watchedRoots;
+        this.watchedHierarchies = watchedHierarchies;
     }
 
-    private Path locationOrFirstExistingAncestor(Path watchableHierarchy) {
-        if (Files.isDirectory(watchableHierarchy)) {
-            return watchableHierarchy;
-        }
-        return SnapshotWatchedDirectoryFinder.findFirstExistingAncestor(watchableHierarchy);
+    public boolean contains(File watchableHierarchy) {
+        return watchedHierarchies.contains(watchableHierarchy.getAbsolutePath());
     }
 
-    /**
-     * Filters out directories whose ancestor is also among the watched directories.
-     */
-    @VisibleForTesting
-    static Set<Path> resolveHierarchiesToWatch(Stream<Path> directories) {
-        Set<Path> hierarchies = new HashSet<>();
-        directories
-            .sorted(Comparator.comparingInt(Path::getNameCount))
-            .filter(path -> {
-                Path parent = path;
-                while (true) {
-                    parent = parent.getParent();
-                    if (parent == null) {
-                        break;
-                    }
-                    if (hierarchies.contains(parent)) {
-                        return false;
-                    }
-                }
-                return true;
-            })
-            .forEach(hierarchies::add);
-        return hierarchies;
+    public Set<File> getWatchedRoots() {
+        return watchedRoots;
     }
 
-    private static class FilterAlreadyCoveredSnapshotsVisitor implements SnapshotHierarchy.SnapshotVisitor {
-        private final SnapshotHierarchy.SnapshotVisitor delegate;
-        private final FileHierarchySet alreadyCoveredSnapshots;
-
-        public FilterAlreadyCoveredSnapshotsVisitor(SnapshotHierarchy.SnapshotVisitor delegate, FileHierarchySet alreadyCoveredSnapshots) {
-            this.delegate = delegate;
-            this.alreadyCoveredSnapshots = alreadyCoveredSnapshots;
-        }
+    public static WatchedHierarchies resolveWatchedHierarchies(WatchableHierarchies watchableHierarchies, SnapshotHierarchy vfsRoot) {
+        FileHierarchySet watchedHierarchies = DefaultFileHierarchySet.of();
+        for (File watchableHierarchy : watchableHierarchies.getRecentlyUsedHierarchies()) {
+            String watchableHierarchyPath = watchableHierarchy.getAbsolutePath();
+            if (watchedHierarchies.contains(watchableHierarchyPath)) {
+                continue;
+            }
 
-        @Override
-        public void visitSnapshotRoot(FileSystemLocationSnapshot snapshot) {
-            if (!alreadyCoveredSnapshots.contains(snapshot.getAbsolutePath())) {
-                delegate.visitSnapshotRoot(snapshot);
+            if (hasNoContent(vfsRoot.rootSnapshotsUnder(watchableHierarchyPath), watchableHierarchies, watchedHierarchies)) {
```

Comment:
For the non-hierarchical case we need to make sure that directories in hierarchies with no content also are not watched. For missing snapshots, we are trying to watch the first existing parent directory: https://github.com/gradle/gradle/blob/3403a1f87dedb0af241e215b72d80f8bb49a41c8/subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/NonHierarchicalFileWatcherUpdater.java#L73

So I suppose we would need to make sure that those watched directories are only in watched hierarchies, right? If they aren't, then we won't probe the root directory.

Reply:
I'm not sure about this one. Maybe we should review this part together?

Reply:
Do you mean that `SnapshotWatchedDirectoryFinder` should take into account the `WatchableHierarchies`, and not allow watching a parent if it's not under a watchable hierarchy? That makes sense. Would simply not returning anything work here, because we would drop the missing stuff from VFS at the end of the build?

Reply:
OK, this should be fixed now.